File: src/main/java/org/example/eshop/controller/AdminCatalogController.java
package org.example.eshop.controller;

import jakarta.validation.Valid;
import org.example.eshop.dto.CreateLotRequest;
import org.example.eshop.dto.CreateProductRequest;
import org.example.eshop.dto.CreateVariantRequest;
import org.example.eshop.dto.UpdateLotRequest;
import org.example.eshop.dto.UpdateProductRequest;
import org.example.eshop.dto.UpdateVariantRequest;
import org.example.eshop.service.AdminCatalogService;
import org.example.eshop.service.AuditLogService;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/admin")
public class AdminCatalogController {

    private final AdminCatalogService adminCatalogService;
    private final AuditLogService auditLogService;

    public AdminCatalogController(AdminCatalogService adminCatalogService, AuditLogService auditLogService) {
        this.adminCatalogService = adminCatalogService;
        this.auditLogService = auditLogService;
    }

    @PostMapping("/products")
    public ResponseEntity<?> createProduct(@Valid @RequestBody CreateProductRequest request) {
        try {
            var product = adminCatalogService.createProduct(request);
            auditLogService.logAdminAction("CREATE", "Product", product.getId(), "slug=" + request.getSlug());
            return ResponseEntity.status(HttpStatus.CREATED).body(product);
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(java.util.Map.of("error", e.getMessage()));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(java.util.Map.of("error", "Failed to create product"));
        }
    }

    @PostMapping("/variants")
    public ResponseEntity<?> createVariant(@Valid @RequestBody CreateVariantRequest request) {
        try {
            var variant = adminCatalogService.createVariant(request);
            auditLogService.logAdminAction("CREATE", "Variant", variant.getId(), "sku=" + request.getSku() + ", productId=" + request.getProductId());
            return ResponseEntity.status(HttpStatus.CREATED).body(variant);
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(java.util.Map.of("error", e.getMessage()));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(java.util.Map.of("error", "Failed to create variant"));
        }
    }

    @PostMapping("/lots")
    public ResponseEntity<?> createLot(@Valid @RequestBody CreateLotRequest request) {
        try {
            var lot = adminCatalogService.createLot(request);
            auditLogService.logAdminAction("CREATE", "Lot", lot.getId(), "productId=" + request.getProductId() + ", harvestYear=" + request.getHarvestYear());
            return ResponseEntity.status(HttpStatus.CREATED).body(lot);
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(java.util.Map.of("error", e.getMessage()));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(java.util.Map.of("error", "Failed to create lot"));
        }
    }

    @PatchMapping("/products/{id}")
    public ResponseEntity<?> updateProduct(
            @PathVariable Long id,
            @Valid @RequestBody UpdateProductRequest request
    ) {
        try {
            var product = adminCatalogService.updateProduct(id, request);
            auditLogService.logAdminAction("UPDATE", "Product", id, "slug=" + product.getSlug());
            return ResponseEntity.ok(product);
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(java.util.Map.of("error", e.getMessage()));
        } catch (java.util.NoSuchElementException e) {
            return ResponseEntity.notFound().build();
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(java.util.Map.of("error", "Failed to update product"));
        }
    }

    @PatchMapping("/variants/{id}")
    public ResponseEntity<?> updateVariant(
            @PathVariable Long id,
            @Valid @RequestBody UpdateVariantRequest request
    ) {
        try {
            var variant = adminCatalogService.updateVariant(id, request);
            auditLogService.logAdminAction("UPDATE", "Variant", id, "sku=" + variant.getSku());
            return ResponseEntity.ok(variant);
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(java.util.Map.of("error", e.getMessage()));
        } catch (java.util.NoSuchElementException e) {
            return ResponseEntity.notFound().build();
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(java.util.Map.of("error", "Failed to update variant"));
        }
    }

    @PatchMapping("/lots/{id}")
    public ResponseEntity<?> updateLot(
            @PathVariable Long id,
            @Valid @RequestBody UpdateLotRequest request
    ) {
        try {
            var lot = adminCatalogService.updateLot(id, request);
            auditLogService.logAdminAction("UPDATE", "Lot", id, "harvestYear=" + lot.getHarvestYear());
            return ResponseEntity.ok(lot);
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(java.util.Map.of("error", e.getMessage()));
        } catch (java.util.NoSuchElementException e) {
            return ResponseEntity.notFound().build();
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(java.util.Map.of("error", "Failed to update lot"));
        }
    }

    @DeleteMapping("/products/{id}")
    public ResponseEntity<?> deleteProduct(@PathVariable Long id) {
        try {
            adminCatalogService.deleteProduct(id);
            auditLogService.logAdminAction("DELETE", "Product", id, null);
            return ResponseEntity.noContent().build();
        } catch (IllegalStateException e) {
            return ResponseEntity.status(HttpStatus.CONFLICT).body(java.util.Map.of("error", e.getMessage()));
        } catch (java.util.NoSuchElementException e) {
            return ResponseEntity.notFound().build();
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(java.util.Map.of("error", "Failed to delete product"));
        }
    }

    @DeleteMapping("/variants/{id}")
    public ResponseEntity<?> deleteVariant(@PathVariable Long id) {
        try {
            adminCatalogService.deleteVariant(id);
            auditLogService.logAdminAction("DELETE", "Variant", id, null);
            return ResponseEntity.noContent().build();
        } catch (IllegalStateException e) {
            return ResponseEntity.status(HttpStatus.CONFLICT).body(java.util.Map.of("error", e.getMessage()));
        } catch (java.util.NoSuchElementException e) {
            return ResponseEntity.notFound().build();
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(java.util.Map.of("error", "Failed to delete variant"));
        }
    }

    @DeleteMapping("/lots/{id}")
    public ResponseEntity<?> deleteLot(@PathVariable Long id) {
        try {
            adminCatalogService.deleteLot(id);
            auditLogService.logAdminAction("DELETE", "Lot", id, null);
            return ResponseEntity.noContent().build();
        } catch (IllegalStateException e) {
            return ResponseEntity.status(HttpStatus.CONFLICT).body(java.util.Map.of("error", e.getMessage()));
        } catch (java.util.NoSuchElementException e) {
            return ResponseEntity.notFound().build();
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(java.util.Map.of("error", "Failed to delete lot"));
        }
    }
}

File: src/main/kotlin/org/example/eshop/controller/AdminCatalogController.kt
package org.example.eshop.controller

import org.example.eshop.dto.CreateProductRequest
import org.example.eshop.dto.CreateVariantRequest
import org.example.eshop.dto.CreateLotRequest
import org.example.eshop.dto.UpdateProductRequest
import org.example.eshop.dto.UpdateVariantRequest
import org.example.eshop.dto.UpdateLotRequest
import org.example.eshop.service.AdminCatalogService
import org.example.eshop.service.AuditLogService
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import jakarta.validation.Valid

@RestController
@RequestMapping("/api/admin")
class AdminCatalogController(
    private val adminCatalogService: AdminCatalogService,
    private val auditLogService: AuditLogService
) {

    @PostMapping("/products")
    fun createProduct(@Valid @RequestBody request: CreateProductRequest): ResponseEntity<Any> {
        return try {
            val product = adminCatalogService.createProduct(request)
            auditLogService.logAdminAction("CREATE", "Product", product.id, "slug=${request.slug}")
            ResponseEntity.status(HttpStatus.CREATED).body(product)
        } catch (e: IllegalArgumentException) {
            ResponseEntity.badRequest().body(mapOf("error" to e.message))
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(mapOf("error" to "Failed to create product"))
        }
    }

    @PostMapping("/variants")
    fun createVariant(@Valid @RequestBody request: CreateVariantRequest): ResponseEntity<Any> {
        return try {
            val variant = adminCatalogService.createVariant(request)
            auditLogService.logAdminAction("CREATE", "Variant", variant.id, "sku=${request.sku}, productId=${request.productId}")
            ResponseEntity.status(HttpStatus.CREATED).body(variant)
        } catch (e: IllegalArgumentException) {
            ResponseEntity.badRequest().body(mapOf("error" to e.message))
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(mapOf("error" to "Failed to create variant"))
        }
    }

    @PostMapping("/lots")
    fun createLot(@Valid @RequestBody request: CreateLotRequest): ResponseEntity<Any> {
        return try {
            val lot = adminCatalogService.createLot(request)
            auditLogService.logAdminAction("CREATE", "Lot", lot.id, "productId=${request.productId}, harvestYear=${request.harvestYear}")
            ResponseEntity.status(HttpStatus.CREATED).body(lot)
        } catch (e: IllegalArgumentException) {
            ResponseEntity.badRequest().body(mapOf("error" to e.message))
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(mapOf("error" to "Failed to create lot"))
        }
    }

    @PatchMapping("/products/{id}")
    fun updateProduct(
        @PathVariable id: Long,
        @Valid @RequestBody request: UpdateProductRequest
    ): ResponseEntity<Any> {
        return try {
            val product = adminCatalogService.updateProduct(id, request)
            auditLogService.logAdminAction("UPDATE", "Product", id, "slug=${product.slug}")
            ResponseEntity.ok(product)
        } catch (e: IllegalArgumentException) {
            ResponseEntity.badRequest().body(mapOf("error" to e.message))
        } catch (e: NoSuchElementException) {
            ResponseEntity.notFound().build()
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(mapOf("error" to "Failed to update product"))
        }
    }

    @PatchMapping("/variants/{id}")
    fun updateVariant(
        @PathVariable id: Long,
        @Valid @RequestBody request: UpdateVariantRequest
    ): ResponseEntity<Any> {
        return try {
            val variant = adminCatalogService.updateVariant(id, request)
            auditLogService.logAdminAction("UPDATE", "Variant", id, "sku=${variant.sku}")
            ResponseEntity.ok(variant)
        } catch (e: IllegalArgumentException) {
            ResponseEntity.badRequest().body(mapOf("error" to e.message))
        } catch (e: NoSuchElementException) {
            ResponseEntity.notFound().build()
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(mapOf("error" to "Failed to update variant"))
        }
    }

    @PatchMapping("/lots/{id}")
    fun updateLot(
        @PathVariable id: Long,
        @Valid @RequestBody request: UpdateLotRequest
    ): ResponseEntity<Any> {
        return try {
            val lot = adminCatalogService.updateLot(id, request)
            auditLogService.logAdminAction("UPDATE", "Lot", id, "harvestYear=${lot.harvestYear}")
            ResponseEntity.ok(lot)
        } catch (e: IllegalArgumentException) {
            ResponseEntity.badRequest().body(mapOf("error" to e.message))
        } catch (e: NoSuchElementException) {
            ResponseEntity.notFound().build()
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(mapOf("error" to "Failed to update lot"))
        }
    }

    @DeleteMapping("/products/{id}")
    fun deleteProduct(@PathVariable id: Long): ResponseEntity<Any> {
        return try {
            adminCatalogService.deleteProduct(id)
            auditLogService.logAdminAction("DELETE", "Product", id)
            ResponseEntity.noContent().build()
        } catch (e: IllegalStateException) {
            ResponseEntity.status(HttpStatus.CONFLICT).body(mapOf("error" to e.message))
        } catch (e: NoSuchElementException) {
            ResponseEntity.notFound().build()
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(mapOf("error" to "Failed to delete product"))
        }
    }

    @DeleteMapping("/variants/{id}")
    fun deleteVariant(@PathVariable id: Long): ResponseEntity<Any> {
        return try {
            adminCatalogService.deleteVariant(id)
            auditLogService.logAdminAction("DELETE", "Variant", id)
            ResponseEntity.noContent().build()
        } catch (e: IllegalStateException) {
            ResponseEntity.status(HttpStatus.CONFLICT).body(mapOf("error" to e.message))
        } catch (e: NoSuchElementException) {
            ResponseEntity.notFound().build()
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(mapOf("error" to "Failed to delete variant"))
        }
    }

    @DeleteMapping("/lots/{id}")
    fun deleteLot(@PathVariable id: Long): ResponseEntity<Any> {
        return try {
            adminCatalogService.deleteLot(id)
            auditLogService.logAdminAction("DELETE", "Lot", id)
            ResponseEntity.noContent().build()
        } catch (e: IllegalStateException) {
            ResponseEntity.status(HttpStatus.CONFLICT).body(mapOf("error" to e.message))
        } catch (e: NoSuchElementException) {
            ResponseEntity.notFound().build()
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(mapOf("error" to "Failed to delete lot"))
        }
    }
}